### PR TITLE
Optimize constant label pair adding with relabel.Replace

### PR DESCRIPTION
**Pull request by damnever**
```
name                       old time/op    new time/op    delta
Relabel/example-8            2.30µs ±69%    1.49µs ±26%  -35.20%  (p=0.000 n=20+20)
Relabel/kubernetes-8         6.02µs ± 5%    5.87µs ± 4%   -2.39%  (p=0.000 n=18+20)
Relabel_ReplaceAddLabel-8     973ns ±16%     546ns ±24%  -43.91%  (p=0.000 n=20+20)

name                       old alloc/op   new alloc/op   delta
Relabel/example-8              633B ± 4%      574B ± 4%   -9.24%  (p=0.000 n=20+20)
Relabel/kubernetes-8         1.46kB ± 1%    1.38kB ± 2%   -5.37%  (p=0.000 n=20+20)
Relabel_ReplaceAddLabel-8    1.81kB ± 0%    1.55kB ± 0%  -14.22%  (p=0.000 n=20+20)

name                       old allocs/op  new allocs/op  delta
Relabel/example-8              17.0 ± 0%      12.0 ± 0%  -29.41%  (p=0.000 n=20+20)
Relabel/kubernetes-8           36.0 ± 0%      31.0 ± 0%  -13.89%  (p=0.000 n=20+20)
Relabel_ReplaceAddLabel-8      19.0 ± 0%       4.0 ± 0%  -78.95%  (p=0.000 n=20+20)
```

### CODE REVIEW

**Comment by bboreham** [comment]
I'm not sure I understand.  What is the significance of looking for `$` ?

File: model/relabel/relabel.go [hunk 1]
```diff
@@ -255,6 +255,17 @@ func relabel(cfg *Config, lb *labels.Builder) (keep bool) {
 			return false
 		}
 	case Replace:
+		// Fast path to add label pair.
```
Comment (bboreham) [inline model/relabel/relabel.go:258]: Maybe this should say "add or delete", since one path leads to `Del()` ?

**Comment by roidelapluie** [comment]
This is not correct as the default relabel config does not replace label values. It only replaces label values that do not have a new line in it.

**Comment by damnever** [comment]
This PR is optimizing the cases like this:
```yaml
- action: replace
  target:  contains_no_variables
  replacement:  contains_no_variables
  separator: ;    // the default
  regex: (.*)     // the default
```
this add an new label pair: `contains_no_variables=contains_no_variables`.

If the regex matches all and there is no variables (the symbol `$`) in the templates (the target and replacement), then we can skip `Regex.FindStringSubmatchIndex` and `Regex.ExpandString`.

File: model/relabel/relabel.go (hunk 1 above)
Comment (damnever) [reply model/relabel/relabel.go:258]: Yes.. this logic is a copy from the below, the invalid label is dropped.

**Comment by damnever** [comment]
This is merely a quick hack, and there might be a more prevalent scenario that we could address.

**Comment by damnever** [comment]
> What is the significance of looking for `$` ?

I think you might be right, there is no need to look for `$`.

**Comment by roidelapluie** [comment]
> This PR is optimizing the cases like this:
> 
> ```yaml
> - action: replace
>   target:  contains_no_variables
>   replacement:  contains_no_variables
>   separator: ;    // the default
>   regex: (.*)     // the default
> ```
> 
> this add an new label pair: `contains_no_variables=contains_no_variables`.
> 
> If the regex matches all and there is no variables (the symbol `$`) in the templates (the target and replacement), then we can skip `Regex.FindStringSubmatchIndex` and `Regex.ExpandString`.

It only adds a label it the target label down not contain new lines

**Comment by bboreham** [comment]
> does not contain new lines 

(Might be confusing: this is the result of Prometheus evaluating regexps in multi-line mode. Which is arguably a bug, but not one we can fix without a feature flag or major version bump.)

**Comment by roidelapluie** [comment]
Yes, this is something we plan to fix in Prometheus 3.

**Comment by damnever** [comment]
@roidelapluie [this commit check the new line in the target](https://github.com/prometheus/prometheus/pull/12180/commits/3d981c8e2cedf19f1339b626252d9eed2453a1c5)

**Comment by damnever** [comment]
@roidelapluie I believe that after a successful replacement, the invalid label will be dropped https://github.com/prometheus/prometheus/blob/fe639b545fa8564cf3ea57a7a0b815e432ab7aed/model/relabel/relabel.go#L263-L267

File: model/relabel/relabel.go [hunk 2]
```diff
@@ -255,6 +255,18 @@ func relabel(cfg *Config, lb *labels.Builder) (keep bool) {
 			return false
 		}
 	case Replace:
+		// Fast path to add or delete label pair.
+		if val == "" && cfg.Regex == DefaultRelabelConfig.Regex &&
+			!containsNewLine(cfg.TargetLabel) &&
```
Comment (roidelapluie) [inline model/relabel/relabel.go:260]: the new line can be in the value, not in the label name.

**Comment by roidelapluie** [comment]
We have looked at this pull request during our bug scrub.

I have added this to the 3.0 milestone where we will fix the default regex, making this pull request simpler and more valuable.


Thank you for your contribution.

**Comment by bboreham** [comment]
> This PR is optimizing the cases like this:

> 
```
- action: replace
  target:  contains_no_variables
  replacement:  contains_no_variables
  separator: ;    // the default
  regex: (.*)     // the default
```

If you add one more condition:
```
source_labels: // empty (the default)
```

then the whole discussion about newlines goes away. (It was already moot to pass `if val == ""` in your code).

Also `lb.Set` already does `Del` if the replacement is blank so you don't need to check that.
In #13308 I propose a check that `TargetLabel` is valid, in the case it does not contain `$`.

All the above lets the optimization collapse to:
```
		// Fast path to add static label pair.
		if val == "" && cfg.Regex == DefaultRelabelConfig.Regex &&
			!varInRegexTemplate(cfg.TargetLabel) && !varInRegexTemplate(cfg.Replacement) {
			lb.Set(cfg.TargetLabel, cfg.Replacement)
			break
		}
```

I don't think we need a completely new Benchmark function; it should be possible to do your focus as a new case within the existing `BenchmarkRelabel`.

**Comment by damnever** [comment]
@bboreham great explanation, I think I was somewhat confused before.

After some experiments, I believe the SourceLabels don't really matter either, since the default regex is `(.*)`. Therefore, it seems okay to remove the `val == ""` check as well.

**Comment by bboreham** [comment]
Please don't do that; we will be back into the newline discussion.
`val == ""` guarantees we don't have to talk about newlines.
`(.*)` does not match `"\n"`, for instance.

**Comment by bboreham** [comment]
My position is this PR does not need to talk about newlines, therefore let's not.

**Review by krajorama — APPROVED** [review]
LGTM

**Review by bboreham — APPROVED** [review]
Thanks!

**Comment by bboreham** [comment]
BTW this should not have been flagged as a 3.0 change; it should be invisible to users and a pure optimisation.

**Pull request merged**